gpflow/kernels/changepoints.py

```diff
@@ -0,0 +1,99 @@
+from collections.abc import Iterable
+from typing import List, Union, Optional
+
+import tensorflow as tf
+
+from .base import Combination, Kernel
+from ..base import Parameter, positive
+
+
+class ChangePoints(Combination):
+    r"""
+    The ChangePoints kernel defines a fixed number of change-points along a 1d
+    input space where different kernels govern different parts of the space.
+
+    The kernel is by multiplication and addition of the base kernels with
+    sigmoid functions (σ). A single change-point kernel is defined as:
+
+        K₁(x, x') * (1 - σ(x)) * (1 - σ(x')) + K₂(x, x') * σ(x) * σ(x')
+
+    where K₁ is deactivated around the change-point and K₂ is activated. The
+    single change-point version can be found in \citet{lloyd2014}. Each sigmoid
+    is a logistic function defined as:
+
+        σ(x) = 1 / (1 + exp{-s(x - x₀)})
+
+    parameterized by location "x₀" and steepness "s".
+
+    @incollection{lloyd2014,
+      author = {Lloyd, James Robert et al},
+      title = {Automatic Construction and Natural-language Description of Nonparametric Regression Models},
+      booktitle = {Proceedings of the Twenty-Eighth AAAI Conference on Artificial Intelligence},
+      year = {2014},
+      url = {http://dl.acm.org/citation.cfm?id=2893873.2894066},
+    }
+    """
+    def __init__(self,
+                 kernels: List[Kernel],
+                 locations: List[float],
+                 steepness: Union[float, List[float]] = 1.0,
+                 name: Optional[str] = None):
+        """
+        :param kernels: list of kernels defining the different regimes
+        :param locations: list of change-point locations in the 1d input space
+        :param steepness: the steepness parameter(s) of the sigmoids, this can be
+            common between them or decoupled
+        """
+        if len(kernels) != len(locations) + 1:
+            raise ValueError("Number of kernels ({nk}) must be one more than the number of "
+                             "changepoint locations ({nl})".format(nk=len(kernels), nl=len(locations)))
+
+        if isinstance(steepness, Iterable) and len(steepness) != len(locations):
+            raise ValueError("Dimension of steepness ({ns}) does not match number of changepoint "
+                             "locations ({nl})".format(ns=len(steepness), nl=len(locations)))
+
+        super().__init__(kernels, name=name)
+
+        self.locations = Parameter(locations)
+        self.steepness = Parameter(steepness, transform=positive())
+
+    def _set_kernels(self, kernels):
+        # it is not clear how to flatten out nested change-points
+        self.kernels = list(kernels)
+
+    def K(self, X, X2=None, presliced=False):
```

Comment:
```suggestion
    def K(self, X: tf.Tensor, X2: Optional[tf.Tensor] = None, presliced: bool = False) -> tf.Tensor:
```

^^^ this also might require to import `Optional` from typing